### DISCUSSION TITLE
Timestamp rounding bug?

### DISCUSSION THREAD

**Discussion by Grumium** [arendst/Tasmota#22349] [Support request]
Hi,
I am simply using the "Time 4" command to obtain a timestamp with millisecond resolution.
`tasmota.cmd("Time 4",true)["Time"]`
I take this string and I am publishing it via MQTT. Below, I posted a snippet of the output in the web console to showcase my problem. Please take a look at the timestamps printed in bold. I am running Tasmota 14.2.0 on an ESP32.
I doubt this is intentional, but I wanted to discuss with you first, before opening an issue. 
Thank you for your support!

> 
> 16:05:45.041 MQT: test/nrt_tele/SENSOR = {"Time":"2024-10-23T16:05:45.027"}
> 16:05:46.017 MQT: test/nrt_tele/SENSOR = {"Time":"2024-10-23T16:05:46.004"}
> 16:05:47.018 MQT: test/nrt_tele/SENSOR = {"Time":"2024-10-23T16:05:47.003"}
> **16:05:48.013 MQT: test/nrt_tele/SENSOR = {"Time":"2024-10-23T16:05:47.000"}**
> 16:05:49.009 MQT: test/nrt_tele/SENSOR = {"Time":"2024-10-23T16:05:48.996"}
> 16:05:50.007 MQT: test/nrt_tele/SENSOR = {"Time":"2024-10-23T16:05:49.993"}
> (...)
> 16:06:00.015 MQT: test/nrt_tele/SENSOR = {"Time":"2024-10-23T16:06:00.002"}
> 16:06:01.010 MQT: test/nrt_tele/SENSOR = {"Time":"2024-10-23T16:06:00.996"}
> 16:06:02.019 MQT: test/nrt_tele/SENSOR = {"Time":"2024-10-23T16:06:02.006"}
> 16:06:03.025 MQT: test/nrt_tele/SENSOR = {"Time":"2024-10-23T16:06:03.012"}
> 16:06:04.272 MQT: test/nrt_tele/SENSOR = {"Time":"2024-10-23T16:06:04.258"}
> 16:06:05.021 MQT: test/nrt_tele/SENSOR = {"Time":"2024-10-23T16:06:05.009"}
> 16:06:06.022 MQT: test/nrt_tele/SENSOR = {"Time":"2024-10-23T16:06:06.009"}
> **16:06:07.014 MQT: test/nrt_tele/SENSOR = {"Time":"2024-10-23T16:06:06.000"}**
> 16:06:08.020 MQT: test/nrt_tele/SENSOR = {"Time":"2024-10-23T16:06:08.008"}

**Comment by s-hadinger** [comment]
Nice one. I'm surprised it happens so often. Are you doing the command at specific time? Using cron?

**Reply by Grumium** [reply]
I am triggering it in a berry driver with `every_second()`.
I do not trigger it at specific times. It's just coincidence that it sometimes hits the exact millisecond.

**Comment by arendst** [comment]
It's a simple "bug/feature". As time is is second based I had to add a seperate milisecond counter. This one is hard to sync with the second interrupt without forced missed milliseconds. If you find a beter way, pls provide one. The code is in support_rtc.ino

**Reply by s-hadinger** [reply]
Not an easy one. Maybe we can minimize desync between time and millis by reordering reads:


From
```C
  String dt = GetDT(time);  // 2017-03-07T11:08:02

  if (DT_LOCAL_MILLIS == time_type) {
    char ms[10];
    snprintf_P(ms, sizeof(ms), PSTR(".%03d"), RtcMillis());
    dt += ms;               // 2017-03-07T11:08:02.123
    time_type = DT_LOCAL;
  }
```

To (time and time_millis are frozen before the calls to string functions that can take some time)
```C
  uint32_t time_millis = RtcMillis();
  String dt = GetDT(time);  // 2017-03-07T11:08:02

  if (DT_LOCAL_MILLIS == time_type) {
    char ms[10];
    snprintf_P(ms, sizeof(ms), PSTR(".%03d"), time_millis);
    dt += ms;               // 2017-03-07T11:08:02.123
    time_type = DT_LOCAL;
  }
```

Other option would be to test if `RtcMillis()` overflows and add 1 second to `time` in such case

**Reply by Grumium** [reply]
Calculating time_millis before dt is probably a good idea. I'll give it a shot.
Following your suggestion but leaving time_millis inside the if-statement, what do you think about...
```C
  if (DT_LOCAL_MILLIS == time_type) {
    char ms[10];
    uint32_t time_millis = millis() - Rtc.millis;
    if (time_millis >= 1000) {
      time += 1
    }
    String dt = GetDT(time);  // 2017-03-07T11:08:02(+1)
    snprintf_P(ms, sizeof(ms), PSTR(".%03d"), time_millis % 1000);
    dt += ms;               // 2017-03-07T11:08:02.123
    time_type = DT_LOCAL;
  } 
  else {
    String dt = GetDT(time);  // 2017-03-07T11:08:02
  }
```
and removing the `RtcMillis()` function?
Haven't tested it yet.

**Reply by s-hadinger** [reply]
That was an option as well. But be aware that the Tasmota loop may have been frozen for more than a full second (even if it's a bad practice). So I would narrow the test for millis between 1000 and 1010.

**Reply by Grumium** [reply]
Hi, unfortunately, your first suggestion
```C
  uint32_t time_millis = RtcMillis();
  String dt = GetDT(time);  // 2017-03-07T11:08:02
```
does not do the trick. I will try some more things.
Here some more examples from an ESP32C3 I have at home.
The "12:04:32.003" and "12:04:33.003" inconsistency is worrying me a little. 

> 12:04:30.998 2024-10-25T12:04:30.996
> 12:04:32.003 **2024-10-25T12:04:31.000 ← TIme 4 wrong**
> 12:04:33.003 **2024-10-25T12:04:33.000 ← TIme 4 correct!?**
> (...)
> 12:07:05.997 2024-10-25T12:07:05.994
> 12:07:07.003 **2024-10-25T12:07:06.000**  ← **TIme 4**
> 12:07:08.002 2024-10-25T12:07:07.999
> (...)
> 12:07:51.998 2024-10-25T12:07:51.995
> 12:07:53.003 **2024-10-25T12:07:52.000**  ← **TIme 4**
> 12:07:54.004 2024-10-25T12:07:54.001
> (...)
> 12:16:51.995 2024-10-25T12:16:51.992
> **12:16:52.000** 2024-10-25T12:16:52.997 ← **system time**
> **12:16:53.000** 2024-10-25T12:16:53.997 ← **system time**
> 12:16:54.955 2024-10-25T12:16:54.952
> (...)
> 12:25:16.990 2024-10-25T12:25:16.980 
> 12:25:18.011 **2024-10-25T12:25:17.000** ← **TIme 4**
> 12:25:18.995 2024-10-25T12:25:18.985 
> **12:25:19.000** 2024-10-25T12:25:19.990 ← **system time**
> **12:25:20.000** 2024-10-25T12:25:20.990 ← **system time**
> 12:25:22.006 2024-10-25T12:25:21.995

**Reply by Grumium** [reply]
The following code solves the problem for ~90% of the cases

```C
String GetDateAndTime(uint8_t time_type) {
  // "2017-03-07T11:08:02-07:00" - ISO8601:2004
  uint32_t time_millis = RtcMillis();
  uint32_t time = Rtc.local_time;
  switch (time_type) {
    case DT_UTC:
      time = Rtc.utc_time;
      break;
//    case DT_LOCALNOTZ:  // Is default anyway but allows for not appendig timezone
//      time = Rtc.local_time;
//      break;
    case DT_DST:
      time = Rtc.daylight_saving_time;
      break;
    case DT_STD:
      time = Rtc.standard_time;
      break;
    case DT_RESTART:
      if (Rtc.restart_time == 0) {
        return "";
      }
      time = Rtc.restart_time;
      break;
    case DT_BOOTCOUNT:
      time = Settings->bootcount_reset_time;
      break;
  }
  String dt = GetDT(time);  // 2017-03-07T11:08:02

  if (DT_LOCAL_MILLIS == time_type) {
    char ms[10];
    if (time_millis == 0) {
      dt = GetDT(time+1);
    }
    snprintf_P(ms, sizeof(ms), PSTR(".%03d"), time_millis);
    dt += ms;               // 2017-03-07T11:08:02.123
    time_type = DT_LOCAL;
  } 
``` 

However, at the position of the red arrow, it still didn't work,
![Untitled](https://github.com/user-attachments/assets/5cc7f576-c918-4bce-b32c-8710960e0001)

 and the system timestamps are also not correct, but this handled elsewhere, I think:
https://github.com/arendst/Tasmota/blob/0f2b3b189880dc6ff9474c0abb6874b39611196d/tasmota/tasmota_support/support.ino#L2603-L2604

**Reply by Jason2866** [reply]
The code size increase should kept small. It is just an information time stamp.

**Reply by Grumium** [reply]
> The code size increase should kept small. It is just an information time stamp.

For the log, I agree, but for sensor data, timestamp duplicates and not monotonically increasing timestamps are not just a cosmetic flaw. And so far it is just one additional variable. There is no major increase in code size.

As a workaround to all of this, I will probably delay the start of my driver with `tasmota.cmd("delay -1",true)` and set a timer with ~100 ms for the subsequent tasks so that I am always in the safe zone with the timestamp.

```
def every_second()
    tasmota.set_timer(100, /-> <func> )
end
```
But I still hope this will be solved.

**Reply by s-hadinger** [reply]
Thanks. Let me have a look

**Reply by Grumium** [reply]
Thank you!

I did spend quite some time now on investigating this. Actually, minimizing the desync is not really advantageous, I am afraid. Surprisingly, I am getting the same 90% success rate with the following adjustment only.
```C
  if (DT_LOCAL_MILLIS == time_type) {
    char ms[10];
    uint32_t time_millis = RtcMillis();
    if (time_millis == 0) {
      dt = GetDT(time+1);
    }
    snprintf_P(ms, sizeof(ms), PSTR(".%03d"), time_millis);
    dt += ms;               // 2017-03-07T11:08:02.123
    time_type = DT_LOCAL;
  }
```
I would explain this as follows: `time` is defined above the conditional statements, thus calling `RTCMillis()` and testing for `time_millis == 0` 'as late as possible' yields quite stable results, since we stay away from the tipping point, i.e., when `time` is not longer equal to the true local time.
